src/dbnode/persist/fs/cross_block_reader.go

```diff
@@ -0,0 +1,254 @@
+	xerrors "github.com/m3db/m3/src/x/errors"
+	"github.com/m3db/m3/src/x/ident"
+)
+
+var (
+	errReaderNotOrderedByIndex                = errors.New("CrossBlockReader can only use DataFileSetReaders ordered by index")
+	errEmptyReader                            = errors.New("trying to read from empty reader")
+	_                          heap.Interface = (*minHeap)(nil)
+)
+
+type crossBlockReader struct {
+	dataFileSetReaders []DataFileSetReader
+	id                 ident.ID
+	tags               ident.TagIterator
+	records            []BlockRecord
+	initialized        bool
+	minHeap            minHeap
+	err                error
+}
+
+// NewCrossBlockReader constructs a new CrossBlockReader based on given DataFileSetReaders.
+// DataFileSetReaders must be configured to return the data in the order of index, and must be
+// provided in a slice sorted by block start time.
+// Callers are responsible for closing the DataFileSetReaders.
+func NewCrossBlockReader(dataFileSetReaders []DataFileSetReader) (CrossBlockReader, error) {
+	var previousStart time.Time
+	for _, dataFileSetReader := range dataFileSetReaders {
+		if !dataFileSetReader.IsOrderedByIndex() {
+			return nil, errReaderNotOrderedByIndex
+		}
+		currentStart := dataFileSetReader.Range().Start
+		if !currentStart.After(previousStart) {
+			return nil, fmt.Errorf("dataFileSetReaders are not ordered by time (%s followed by %s)", previousStart, currentStart)
+		}
+		previousStart = currentStart
+	}
+
+	return &crossBlockReader{
+		dataFileSetReaders: append([]DataFileSetReader{}, dataFileSetReaders...),
+		records:            make([]BlockRecord, 0, len(dataFileSetReaders)),
+	}, nil
+}
+
+func (r *crossBlockReader) Next() bool {
+	if !r.initialized {
+		err := r.init()
+		if err != nil {
+			r.err = err
+			return false
+		}
+	}
+
+	if len(r.minHeap) == 0 {
+		return false
+	}
+
+	firstEntry, err := r.readOne()
+	if err != nil {
+		r.err = err
+		return false
+	}
+
+	r.id = firstEntry.id
+	r.tags = firstEntry.tags
+
+	// use empty var in inner loop with "for i := range" to have compiler use memclr optimization
+	// see: https://codereview.appspot.com/137880043
+	var emptyRecord BlockRecord
+	for i := range r.records {
+		r.records[i] = emptyRecord
+	}
+	r.records = r.records[:0]
+	r.records = append(r.records, BlockRecord{firstEntry.data, firstEntry.checksum})
+
+	for len(r.minHeap) > 0 && r.minHeap[0].id.Equal(firstEntry.id) {
+		nextEntry, err := r.readOne()
+		if err != nil {
+			// Close the resources that were already read but not returned to the consumer:
+			r.id.Finalize()
+			r.tags.Close()
+			for _, record := range r.records {
+				record.Data.DecRef()
+				record.Data.Finalize()
+			}
+			for i := range r.records {
+				r.records[i] = emptyRecord
+			}
+			r.records = r.records[:0]
+			r.err = err
+			return false
+		}
+		nextEntry.id.Finalize()
+		nextEntry.tags.Close()
+		r.records = append(r.records, BlockRecord{nextEntry.data, nextEntry.checksum})
```

Comment:
Good point, just that the `len(dataFileSetReaders)` is not the best indicator for duplicate ids because I invalidate (set to `null`) already exhausted `dataFileSetReaders` and don't track the number of still valid ones. Instead, implemented a more direct check for duplicate ids in `readOne` (and pushed as a separate commit for clarity).
Just not completely sure, for our use case, would it make more sense to return an error and fail the whole process, or to skip the duplicate. What do you think?

Reply:
As per our conversation, keeping the error approach, and reporting it as Invariant violation.